cpp/src/io/parquet/bloom_filter_reader.cu

```diff
@@ -0,0 +1,688 @@
+                                      stream);
+
+          read_tasks.emplace_back(std::move(future_read_size));
+        } else {
+          buffer                   = source->host_read(bitset_offset, bitset_size);
+          bloom_filter_data[chunk] = rmm::device_buffer{buffer->data(), buffer->size(), stream};
+        }
+      }
+    });
+
+  // Read task sync function
+  for (auto& task : read_tasks) {
+    task.wait();
+  }
+}
+
+}  // namespace
+
+std::vector<rmm::device_buffer> aggregate_reader_metadata::read_bloom_filters(
+  host_span<std::unique_ptr<datasource> const> sources,
+  host_span<std::vector<size_type> const> row_group_indices,
+  host_span<int const> column_schemas,
+  size_type total_row_groups,
+  rmm::cuda_stream_view stream) const
+{
+  // Descriptors for all the chunks that make up the selected columns
+  auto const num_input_columns = column_schemas.size();
+  auto const num_chunks        = total_row_groups * num_input_columns;
+
+  // Association between each column chunk and its source
+  std::vector<size_type> chunk_source_map(num_chunks);
+
+  // Keep track of column chunk file offsets
+  std::vector<std::optional<int64_t>> bloom_filter_offsets(num_chunks);
+  std::vector<std::optional<int32_t>> bloom_filter_sizes(num_chunks);
+
+  // Gather all bloom filter offsets and sizes.
+  size_type chunk_count = 0;
+
+  // Flag to check if we have at least one valid bloom filter offset
+  auto have_bloom_filters = false;
+
+  // For all data sources
+  std::for_each(thrust::counting_iterator<size_t>(0),
+                thrust::counting_iterator(row_group_indices.size()),
+                [&](auto const src_index) {
+                  // Get all row group indices in the data source
+                  auto const& rg_indices = row_group_indices[src_index];
+                  // For all row groups
+                  std::for_each(rg_indices.cbegin(), rg_indices.cend(), [&](auto const rg_index) {
+                    // For all column chunks
+                    std::for_each(
+                      column_schemas.begin(), column_schemas.end(), [&](auto const schema_idx) {
+                        auto& col_meta = get_column_metadata(rg_index, src_index, schema_idx);
+
+                        // Get bloom filter offsets and sizes
+                        bloom_filter_offsets[chunk_count] = col_meta.bloom_filter_offset;
+                        bloom_filter_sizes[chunk_count]   = col_meta.bloom_filter_length;
+
+                        // Set `have_bloom_filters` if `bloom_filter_offset` is valid
+                        if (col_meta.bloom_filter_offset.has_value()) { have_bloom_filters = true; }
+
+                        // Map each column chunk to its source index
+                        chunk_source_map[chunk_count] = src_index;
+                        chunk_count++;
+                      });
+                  });
+                });
+
+  // Do we have any bloom filters
+  if (have_bloom_filters) {
```

Comment:
Maybe invert this `if` statement and treat it as an early exit:

```cpp

if (!have_bloom_filters) {
  // Return empty vector
  return {};
}

// Continue with other logic which is no longer indented an extra level

Reply:
All done, thanks for the suggestion!